### PR TITLE
[Backport 2.x] Force merge API supports performing only on primary shards (#11269)

### PR DESCRIPTION
**Pull request by gaobinlong**
### Description
Backport https://github.com/opensearch-project/OpenSearch/pull/11269 to 2.x.

### Related Issues
https://github.com/opensearch-project/OpenSearch/issues/8021

### Check List
- [x] New functionality includes testing.
  - [ ] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by sohami** [comment]
> ❌ Gradle check result for [e9b194d](https://github.com/opensearch-project/OpenSearch/commit/e9b194d3b66b99eb2c233ffba03cfeab301082a9): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/34697/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Failures are related to: https://github.com/opensearch-project/OpenSearch/issues/12593

**Review by sohami — APPROVED** [review]

**Comment by sohami** [comment]
> ❕ Gradle check result for [e9b194d](https://github.com/opensearch-project/OpenSearch/commit/e9b194d3b66b99eb2c233ffba03cfeab301082a9): [UNSTABLE](https://build.ci.opensearch.org/job/gradle-check/34750/)
> 
> * **TEST FAILURES:**
> 
> ```
>       1 org.opensearch.repositories.azure.AzureBlobStoreRepositoryTests.testContainerCreationAndDeletion
>       1 org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=indices.get_field_mapping/20_missing_field/Return empty object if field doesn't exist, but index does}
>       1 org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=indices.forcemerge/20_wait_for_completion/Force merge index with wait_for_completion}
> ```
> 
> Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

@gaobinlong It looks like below test is flaky. Can you check this please ?
```
>       1 org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=indices.forcemerge/20_wait_for_completion/Force merge index with wait_for_completion}
```

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.forcemerge/20_wait_for_completion.yml [hunk 1]
```diff
@@ -27,6 +29,28 @@
   - match: { task.action: "indices:admin/forcemerge" }
   - match: { task.description: "Force-merge indices [test_index], maxSegments[1], onlyExpungeDeletes[false], flush[true]" }
 
+---
+"Force merge index with wait_for_completion after task description changed":
+  - skip:
+      version: " - 2.12.99 "
+      reason: "task description was changed in 2.13.0"
+      features: allowed_warnings
```
Comment (sohami) [inline rest-api-spec/src/main/resources/rest-api-spec/test/indices.forcemerge/20_wait_for_completion.yml:37]: Do we need to add `node_selector` features here as well to skip any nodes <= 2.12.99 ?

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.forcemerge/20_wait_for_completion.yml [hunk 2]
```diff
@@ -4,15 +4,17 @@
   # will return a task immediately and the merge process will run in background.
 
   - skip:
-      version: " - 2.6.99"
-      reason: "only available in 2.7+"
-      features: allowed_warnings
+      version: " - 2.6.99, 2.13.0 - "
+      reason: "wait_for_completion was introduced in in 2.7.0 and task description was changed in 2.13.0"
```
Comment (sohami) [inline rest-api-spec/src/main/resources/rest-api-spec/test/indices.forcemerge/20_wait_for_completion.yml:8]: there is redundant in in the message `... was introduced in in 2.7.0 ... `

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.forcemerge/20_wait_for_completion.yml (hunk 1 above)
Comment (gaobinlong) [reply rest-api-spec/src/main/resources/rest-api-spec/test/indices.forcemerge/20_wait_for_completion.yml]: Yeah, we need that, thanks!

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.forcemerge/20_wait_for_completion.yml (hunk 2 above)
Comment (gaobinlong) [reply rest-api-spec/src/main/resources/rest-api-spec/test/indices.forcemerge/20_wait_for_completion.yml]: Good catch, thank you, I've changed that.

**Comment by gaobinlong** [comment]
> ❕ Gradle check result for [83fb554](https://github.com/opensearch-project/OpenSearch/commit/83fb55479bfcd629bbe471420dbe31da6d6a36a9): [UNSTABLE](https://build.ci.opensearch.org/job/gradle-check/34769/)
> 
> * **TEST FAILURES:**
> 
> ```
>       1 org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=indices.get_field_mapping/20_missing_field/Return empty object if field doesn't exist, but index does}
> ```
> 
> Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

The failed test seems not related to this PR, and there's an [issue](https://github.com/opensearch-project/OpenSearch/issues/2440) tracking it but seems we haven't resolved it yet.

**Review by kkmr — APPROVED** [review]

**Comment by sohami** [comment]
> ❕ Gradle check result for [7d1c255](https://github.com/opensearch-project/OpenSearch/commit/7d1c2553a2a98888f7e1fa99a96b47ed0d47fe76): [UNSTABLE](https://build.ci.opensearch.org/job/gradle-check/34787/)
> 
> * **TEST FAILURES:**
> 
> ```
>       1 org.opensearch.remotestore.SegmentReplicationUsingRemoteStoreIT.testRestartPrimary_NoReplicas
> ```
> 
> Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

Known issue: https://github.com/opensearch-project/OpenSearch/issues/11993

**Comment by sohami** [comment]
@gaobinlong We will need a forward port to change the version checks in main branch. Also probably we can update the yml tests as done in this PR. Can you also create a documentation issue or PR for this new option ?

**Comment by gaobinlong** [comment]
> @gaobinlong We will need a forward port to change the version checks in main branch. Also probably we can update the yml tests as done in this PR. Can you also create a documentation issue or PR for this new option ?

Thanks, I'll create a following PR for the main branch, and both of the related document [issue](https://github.com/opensearch-project/documentation-website/issues/6662) and [PR](https://github.com/opensearch-project/documentation-website/pull/6664) have been created yet.

**Pull request merged**